File: SpotifyChallenges/ZipfSong.py
def add(targetList, newTuple):
	for i in range(len(targetList)):
		if newTuple[1] > targetList[i][1]:
			targetList.insert(i, newTuple)
			return
	targetList.append(newTuple)

[numCases, numChoose] = [int(n) for n in raw_input().split(' ')]
chosen = []
infoIn = raw_input().split(' ')
numPlays, songName = int(infoIn[0]), infoIn[1]
basePlays = numPlays
chosen.append((songName, 1))

for i in range(2, numCases + 1):
	infoIn = raw_input().split(' ')
	numPlays, songName = int(infoIn[0]), infoIn[1]
	zipfScore = float(basePlays) / i
	zipfScore = float(numPlays) / zipfScore
	if len(chosen) < numChoose:
		add(chosen, (songName, zipfScore))
	elif zipfScore > chosen[-1][1]:
		chosen = chosen[:-1]
		add(chosen, (songName, zipfScore))
		
for songTuple in chosen:
	print songTuple[0]

File: SpotifyChallenges/ZipfSong.cpp
#include <string>
#include <vector>
#include <stdio.h>
#include <iostream>

using namespace std;

void add(vector<string> &names, vector<double> &scores, string nameIn, double scoreIn);
int findIndex(vector<double> &scores, double scoreIn);

int main()
{
	unsigned int numCases, numChoose;
	double basePlays, numPlays, zipfScore;
	string songName;
	vector<double> *chosenScores = new vector<double>();
	vector<string> *chosenNames = new vector<string>();

	cin >> numCases;
	cin >> numChoose;
	
	chosenScores->reserve((vector<double>::size_type) numChoose);
	chosenNames->reserve((vector<string>::size_type) numChoose);

	cin >> basePlays;
	cin >> songName;

	if (basePlays == 0)
	{
		basePlays = 1e-12;
	}

	chosenScores->push_back(1.0);
	chosenNames->push_back(songName);

	for (int i = 2; i < numCases + 1; i++)
	{
		cin >> numPlays;
		cin >> songName;

		zipfScore = basePlays / i;
		zipfScore = numPlays / zipfScore;

		if (chosenScores->size() < numChoose)
		{
			add(*chosenNames, *chosenScores, songName, zipfScore);
		}
		else if (zipfScore > chosenScores->back())
		{
			chosenScores->pop_back();
			chosenNames->pop_back();
			add(*chosenNames, *chosenScores, songName, zipfScore);
		}
	}

	for (unsigned int i = 0; i < chosenNames->size(); i++)
	{
		if (i > 0)
		{
			cout << endl;
		}
		cout << chosenNames->at(i);
	}

	return 0;
}

void add(vector<string> &names, vector<double> &scores, string nameIn, double scoreIn)
{
	unsigned int index = findIndex(scores, scoreIn);
	if (index == scores.size())
	{
		names.push_back(nameIn);
		scores.push_back(scoreIn);
	}
	else
	{
		names.insert(names.begin() + index, nameIn);
		scores.insert(scores.begin() + index, scoreIn);
	}
}

int findIndex(vector<double> &scores, double scoreIn)
{
	int index, upper = scores.size() - 1, lower = 0, range = upper - lower;

	while (upper > lower)
	{
		range = upper - lower;
		index = lower + (range / 2);
		if (scoreIn > scores.at(index))
		{
			upper = index;
		}
		else if (scoreIn <= scores.at(index))
		{
			lower = index + 1;
		}
	}

	if (scoreIn > scores.at(upper))
	{
		return upper;
	}
	else
	{
		return upper + 1;
	}
}